### PR TITLE
[Misc] Upgrade to pytorch 2.5

### PR DESCRIPTION
**Pull request by bnellnm**
Upgrade to pytorch 2.5

Requires changes to flash attn: https://github.com/vllm-project/flash-attention/pull/23

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: requirements-cuda.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # Dependencies for NVIDIA GPUs
 ray >= 2.9
 nvidia-ml-py # for pynvml package
-torch == 2.4.0
+torch == 2.5.0
```
Comment (robertgshaw2-neuralmagic) [inline requirements-cuda.txt:7]: Only concern here is now `torch` uses the `12.4 cuda` bindings by default. We might want to update the installation docs (including on the readme) to alert users that they may want to pass `--extra-index-url https://download.pytorch.org/whl/cu121` during installation

**Comment by youkaichao** [comment]
@bnellnm it loos like there are some cmake errors:

```text
[2024-10-22T15:09:27Z] #25 15.41 CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
--
  | [2024-10-22T15:09:27Z] #25 15.41 Please set them or make sure they are set and tested correctly in the CMake files:
  | [2024-10-22T15:09:27Z] #25 15.41 CUDA_CUDA_LIBRARY (ADVANCED)
  | [2024-10-22T15:09:27Z] #25 15.41     linked by target "_moe_C" in directory /workspace
  | [2024-10-22T15:09:27Z] #25 15.41     linked by target "_C" in directory /workspace
  | [2024-10-22T15:09:27Z] #25 15.41     linked by target "vllm_flash_attn_c" in directory /workspace/.deps/vllm-flash-attn-src
  | [2024-10-22T15:09:27Z] #25 15.41
```

the cuda version should not matter that much. I think our current pipeline should still work even if pytorch itself is built against cuda 12.4 .

**Comment by tlrmchlsmth** [comment]
> @bnellnm it loos like there are some cmake errors:
> 
> ```
> [2024-10-22T15:09:27Z] #25 15.41 CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
> --
>   | [2024-10-22T15:09:27Z] #25 15.41 Please set them or make sure they are set and tested correctly in the CMake files:
>   | [2024-10-22T15:09:27Z] #25 15.41 CUDA_CUDA_LIBRARY (ADVANCED)
>   | [2024-10-22T15:09:27Z] #25 15.41     linked by target "_moe_C" in directory /workspace
>   | [2024-10-22T15:09:27Z] #25 15.41     linked by target "_C" in directory /workspace
>   | [2024-10-22T15:09:27Z] #25 15.41     linked by target "vllm_flash_attn_c" in directory /workspace/.deps/vllm-flash-attn-src
>   | [2024-10-22T15:09:27Z] #25 15.41
> ```
> 
> the cuda version should not matter that much. I think our current pipeline should still work even if pytorch itself is built against cuda 12.4 .

Looks related to https://github.com/vllm-project/vllm/pull/8609/

**Review by ProExpertProg — APPROVED** [review]

File: requirements-cuda.txt (hunk 1 above)
Comment (bnellnm) [reply requirements-cuda.txt:7]: @robertgshaw2-neuralmagic Does this mean that there would need to be multiple vllm packages (one for 12.1 and one for 12.4)?  Or should I try to install pytorch 2.5 built with 12.1 (if such a thing exists)?

File: pyproject.toml [hunk 2]
```diff
@@ -6,7 +6,7 @@ requires = [
     "packaging",
     "setuptools>=61",
     "setuptools-scm>=8.0",
-    "torch == 2.4.0",
+    "torch == 2.5.0 --extra-index-url https://download.pytorch.org/whl/cu121",
```
Comment (tlrmchlsmth) [inline pyproject.toml:9]: Given @youkaichao's comment:
>the cuda version should not matter that much. I think our current pipeline should still work even if pytorch itself is built against cuda 12.4 .

We should consider ditching the `--extra-index-url`. Perhaps this should be configurable, but one thing to note is that 2:4 sparse fp8 will require the Pytorch version that's built with 12.4

**Review by tlrmchlsmth — COMMENTED** [review]
The PR looks good but we should quickly come to a consensus on what to do with the CUDA version that pytorch is built against

**Comment by robertgshaw2-neuralmagic** [comment]
> The PR looks good but we should quickly come to a consensus on what to do with the CUDA version that pytorch is built against

Im fine to remove the `--extra-index-url` if we decide to make 12.4 the default for vllm. But we should still ship 12.1 and 11.8 wheels IMO

File: CMakeLists.txt [hunk 3]
```diff
@@ -507,7 +507,7 @@ else()
   FetchContent_Declare(
           vllm-flash-attn
           GIT_REPOSITORY https://github.com/vllm-project/flash-attention.git
-          GIT_TAG 013f0c4fc47e6574060879d9734c1df8c5c273bd
+          GIT_TAG 5259c586c403a4e4d8bf69973c159b40cc346fb9
```
Comment (tlrmchlsmth) [inline CMakeLists.txt:510]: 👍

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by DarkLight1337** [comment]
Please merge from main to fix the CI failures for multi-modal models.

**Review by youkaichao — APPROVED** [review]
excited to see it happen!

**Comment by youkaichao** [comment]
some errors are real:

> cuDNN Frontend error: [cudnn_frontend] Error: No execution plans support the graph.

https://github.com/huggingface/diffusers/issues/9704

fixing by [0068133](https://github.com/vllm-project/vllm/pull/9588/commits/0068133747c608093a0aa421ee26d70ec635552b)

**Comment by youkaichao** [comment]
pytorch 2.5 changes the output slightly:

> [2024-10-27T04:21:27Z] hf:	‘vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n\n- Advantages of using the LLM:\n  - High-fidelity and accurate predictions: The LLM can generate high-quality and context’
> [2024-10-27T04:21:27Z] vllm:	‘vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n\n- Advantages of using the LLM:\n  - High-efficiency and low-latency inference: The LLM provides fast and efficient’

The output is still sensible.

Therefore I changed it to logprobs check instead.

For future reference, we can also change to logprobs check if exact comparison is not feasible while it is not our fault (due to pytorch or huggingface numerical change).

**Comment by fzyzcjy** [comment]
Hi, looking forward to this new support, I wonder when will it be released? Thanks!

**Comment by youkaichao** [comment]
> Hi, looking forward to this new support, I wonder when will it be released? Thanks!

you can install the wheel from main branch to use it directly. see https://docs.vllm.ai/en/latest/getting_started/installation.html#install-the-latest-code

**Comment by fzyzcjy** [comment]
@youkaichao thanks! but is it less stable than a stable version?

**Pull request merged**